python/tvm/runtime/ndarray.py

```diff
@@ -18,132 +18,36 @@
-
-
-def context(dev_type, dev_id=0):
-    """Construct a TVM context with given device type and id.
-
-    Parameters
-    ----------
-    dev_type: int or str
-        The device type mask or name of the device.
-
-    dev_id : int, optional
-        The integer device id
-
-    Returns
-    -------
-    ctx: TVMContext
-        The corresponding context.
-
-    Examples
-    --------
-    Context can be used to create reflection of context by
-    string representation of the device type.
-
-    .. code-block:: python
-
-      assert tvm.context("cpu", 1) == tvm.cpu(1)
-      assert tvm.context("gpu", 0) == tvm.gpu(0)
-      assert tvm.context("cuda", 0) == tvm.gpu(0)
-    """
-    if isinstance(dev_type, string_types):
-        if '-device=micro_dev' in dev_type:
-            dev_type = 'micro_dev'
-        else:
-            dev_type = dev_type.split()[0]
-            if dev_type not in TVMContext.STR2MASK:
-                raise ValueError("Unknown device type %s" % dev_type)
-            dev_type = TVMContext.STR2MASK[dev_type]
-    return TVMContext(dev_type, dev_id)
-
-
-def numpyasarray(np_data):
-    """Return a TVMArray representation of a numpy array.
-    """
-    data = np_data
-    assert data.flags['C_CONTIGUOUS']
-    arr = TVMArray()
-    shape = c_array(tvm_shape_index_t, data.shape)
-    arr.data = data.ctypes.data_as(ctypes.c_void_p)
-    arr.shape = shape
-    arr.strides = None
-    arr.dtype = TVMType(np.dtype(data.dtype).name)
-    arr.ndim = data.ndim
-    # CPU device
-    arr.ctx = context(1, 0)
-    return arr, shape
-
-
-def empty(shape, dtype="float32", ctx=context(1, 0)):
-    """Create an empty array given shape and device
-
-    Parameters
-    ----------
-    shape : tuple of int
-        The shape of the array
-
-    dtype : type or str
-        The data type of the array.
-
-    ctx : TVMContext
-        The context of the array
-
-    Returns
-    -------
-    arr : tvm.nd.NDArray
-        The array tvm supported.
-    """
-    shape = c_array(tvm_shape_index_t, shape)
-    ndim = ctypes.c_int(len(shape))
-    handle = TVMArrayHandle()
-    dtype = TVMType(dtype)
-    check_call(_LIB.TVMArrayAlloc(
-        shape, ndim,
-        ctypes.c_int(dtype.type_code),
-        ctypes.c_int(dtype.bits),
-        ctypes.c_int(dtype.lanes),
-        ctx.device_type,
-        ctx.device_id,
-        ctypes.byref(handle)))
-    return _make_array(handle, False, False)
+    from tvm._ffi._ctypes.ndarray import _set_class_ndarray, _make_array, _from_dlpack
+    from tvm._ffi._ctypes.ndarray import NDArrayBase
 
 
-def from_dlpack(dltensor):
-    """Produce an array from a DLPack tensor without memory copy.
-    Retreives the underlying DLPack tensor's pointer to create an array from the
-    data. Removes the original DLPack tensor's destructor as now the array is
-    responsible for destruction.
+@tvm._ffi.register_object
+class NDArray(NDArrayBase):
+    """Lightweight NDArray class of TVM runtime.
```

Comment:
I think the comments here could use some rewording.

Reply:
Feel free to send a followup PR